R-package/tests/testthat/test_model_compatibility.R

```diff
@@ -72,13 +72,34 @@ test_that("Models from previous versions of XGBoost can be loaded", {
     m <- regmatches(model_file, m)[[1]]
     model_xgb_ver <- m[2]
     name <- m[3]
+    is_rds <- endsWith(model_file, '.rds')
 
-    if (endsWith(model_file, '.rds')) {
-      booster <- readRDS(model_file)
-    } else {
-      booster <- xgb.load(model_file)
+    cpp_warning <- capture.output({
+      # Expect an R warning when a model is loaded from RDS and it was generated by version < 1.1.x
+      if (is_rds && compareVersion(model_xgb_ver, '1.1.1.1') < 0) {
+        booster <- readRDS(model_file)
+        expect_warning(predict(booster, newdata = pred_data))
+        expect_warning(run_booster_check(booster, name))
+      } else {
+        if (is_rds) {
+          booster <- readRDS(model_file)
+        } else {
+          booster <- xgb.load(model_file)
+        }
+        predict(booster, newdata = pred_data)
+        run_booster_check(booster, name)
+      }
+    })
+    if (compareVersion(model_xgb_ver, '1.0.0.0') < 0) {
+      # Expect a C++ warning when a model was generated in version < 1.0.x
+      m <- grepl(paste0('.*Loading model from XGBoost < 1\\.0\\.0, consider saving it again for ',
+                        'improved compatibility.*'), cpp_warning, perl = TRUE)
+      expect_true(length(m) > 0 && all(m))
+    } else if (is_rds && model_xgb_ver == '1.1.1.1') {
```

Comment:
Version is hardcoded here, so for new models this if-else won't be triggered at all?

Reply:
That's also correct, since no C++ warning is currently generated for the models that were generated by version 1.2.0.x.